### PR TITLE
Node: Loki logging

### PR DESCRIPTION
**Pull request by bruce-riley**
This PR adds the ability for the guardian to log to Grafana / Loki cloud service rather than GCP. The guardian can be configured to log to one or the other (or no cloud logging).

To use Loki instead of GCP, set the `--telemetryURL` guardian config parameter.

### CODE REVIEW

File: node/go.mod [hunk 1]
```diff
@@ -3,19 +3,19 @@ module github.com/certusone/wormhole/node
...
+	github.com/cenkalti/backoff/v4 v4.2.0
+	github.com/coreos/go-systemd v0.0.0-20191104093116-d3cd4ed1dbcf
 	github.com/davecgh/go-spew v1.1.1
 	github.com/dgraph-io/badger/v3 v3.2103.1
 	github.com/ethereum/go-ethereum v1.10.21
 	github.com/gagliardetto/solana-go v1.7.1
 	github.com/gorilla/mux v1.8.0
 	github.com/gorilla/websocket v1.5.0 // indirect
-	github.com/grpc-ecosystem/go-grpc-middleware v1.4.0
+	github.com/grpc-ecosystem/go-grpc-middleware v1.3.0
```
Comment (tbjump) [inline node/go.mod:16]: why this downgrade?

**Review by tbjump — CHANGES_REQUESTED** [review]
Please check the `node/go.mod`. It includes a few downgrades which we probably don't want. You may not have picked up https://github.com/wormhole-foundation/wormhole/pull/3170 yet.

File: node/cmd/guardiand/node.go [hunk 2]
```diff
@@ -352,6 +359,8 @@ func init() {
 	telemetryProject = NodeCmd.Flags().String("telemetryProject", defaultTelemetryProject,
 		"Google Cloud Project to use for Telemetry logging")
```
Comment (panoel) [inline node/cmd/guardiand/node.go:360]: I guess this isn't just for the Google Cloud Project anymore.

**Review by panoel — APPROVED** [review]

File: node/pkg/telemetry/loki.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		BatchSize:              10,
+		DropRateLimitedBatches: true,
+		Client:                 config.HTTPClientConfig{},
+		BackoffConfig:          backoff.Config{MinBackoff: 1 * time.Millisecond, MaxBackoff: 2 * time.Millisecond, MaxRetries: 3},
+		ExternalLabels:         lokiflag.LabelSet{},
+		Timeout:                1 * time.Second,
+		// TenantID: We are not using the tenantID.
+	}
+
+	clientMaxLineSize := 1024
```
Comment (SEJeff) [inline node/pkg/telemetry/loki.go:96]: Out of curiosity, did you do any checking on line lengths of existing logs? I'm assuming if we log VAAs, it is going to bust this every single time. A sample VAA I have for testing `worm parse` is 2097 characters alone.

**Review by SEJeff — DISMISSED** [review]
Approved, with a few questions. If you want to blow them off feel free to merge. It would be nice to get some clarity however if you have the time.

**Review by tbjump — APPROVED** [review]

File: node/cmd/guardiand/node.go [hunk 4]
```diff
@@ -927,18 +921,45 @@ func runNode(cmd *cobra.Command, args []string) {
...
-		if err != nil {
-			logger.Fatal("Failed to initialize telemetry", zap.Error(err))
+
+		var tm *telemetry.Telemetry
+		if *telemetryURL != "" {
+			logger.Info("Using Loki telemetry logger",
+				zap.String("publicRpcLogDetail", *publicRpcLogDetailStr),
+				zap.Bool("logPublicRpcToTelemetry", *publicRpcLogToTelemetry))
+
+			tm, err = telemetry.NewLokiCloudLogger(rootCtx, logger, *telemetryURL, "wormhole", *telemetryProject, true, labels)
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/cmd/guardiand/node.go:931]: Changed to use `context.Background()`.

File: node/go.mod (hunk 1 above)
Comment (bruce-riley) [reply node/go.mod]: Thank you, @tbjump for fixing this!

**Comment by tbjump** [comment]
please squash on merge.

File: node/cmd/guardiand/node.go (hunk 2 above)
Comment (bruce-riley) [reply node/cmd/guardiand/node.go:360]: No longer using this parameter in Loki, so this string is now/still correct.

File: node/pkg/telemetry/loki.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		return nil, fmt.Errorf("failed to parse Loki client url: %v", err)
+	}
+
+	cfg := client.Config{
+		URL:                    serverURL,
+		BatchWait:              100 * time.Millisecond,
+		BatchSize:              10,
+		DropRateLimitedBatches: true,
+		Client:                 config.HTTPClientConfig{},
+		BackoffConfig:          backoff.Config{MinBackoff: 1 * time.Millisecond, MaxBackoff: 2 * time.Millisecond, MaxRetries: 3},
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/telemetry/loki.go]: Updated it to use the defaults from promtail, and point to the above file.

File: node/pkg/telemetry/loki.go (hunk 3 above)
Comment (bruce-riley) [reply node/pkg/telemetry/loki.go:99]: We are no longer logging VAAs, so 1024 *should* be big enough. 

Loki does peg Prometheus metrics, so we will need to monitor those to see if things are getting dropped.

**Comment by bruce-riley** [comment]
FYI, here are the metrics that Loki pegs:
https://github.com/grafana/loki/blob/main/clients/pkg/promtail/client/client.go#L71-L127
We will need to monitor these to see if things are getting dropped.

**Review by panoel — APPROVED** [review]

**Review by tbjump — APPROVED** [review]

**Review by evan-gray — APPROVED** [review]
it's good to have options :wink:

**Pull request merged**